Integration tests that save response and checks for changes.

jan-niestadt opened the pull request:
- makes it quicker to add tests
- tests more thoroughly because the entire response is checked

Of course, we would still need to check manually that the saved responses are actually correct. But even if we just monitor for changes and investigate whenever those occur, there's value in that.

@eginez What do you think? Is this the right approach going forward? I'd like to convert most tests to this approach, do you see any issues with that? Can this work with your separate run with different test settings as well?

jan-niestadt commented:
We would probably want to use subdirs per test category, so `saved-responses/hits/`, `saved-responsed/docs/`, etc.

jan-niestadt commented:
One issue is that the sort operation is not completely defined, resulting in different ordering of hits. I think I will add an option to sort by (doc, start, end) as the final sort option to fix this.

eginez commented on test/test/compare-responses.js:
```diff
@@ -0,0 +1,98 @@
+ * @param response response to sanitize
+ * @return response with variable values replaces with fixed ones
+ */
+function sanitizeResponse(response) {
```
I have a very similar function at indices.js:clearKeys(). Do you think is worth merging these 2 functions into one?

eginez replied:
It seems like the appropriate place to put it these functions is actually here

jan-niestadt replied:
Good idea, done!

eginez commented on test/test/compare-responses.js:
```diff
@@ -0,0 +1,98 @@
+ * @param valuesToFix (potentially nested) object structure of values that, when found, should be set to this fixed value.
+ * @return response with variable values replaces with fixed ones
+ */
+function makeResponseValuesFixed(response, valuesToFix) {
```
This can also be merged with `clearKeys()`

eginez commented on test/test/compare-responses.js:
```diff
@@ -0,0 +1,98 @@
+    // Did we have a previous response?
+    const categoryDir = path.resolve(SAVED_RESPONSES_PATH, sanitizeFileName(category));
+    if (!fs.existsSync(categoryDir))
+        fs.mkdirSync(categoryDir);
```
An optional change: I think you can simplify the logic here by creating the file if the `category` directory is not found and return

jan-niestadt replied:
I'm not sure I understand. Wouldn't that mean we'd have the file creation in two places, e.g:

```js
if (dir doesn't exist) {
  fs.mkdirSync(...)
  fs.writeFileSync(...)
} else if (file-in-dir doesn't exist) {
  fs.writeFileSync(...)
}
```

Of course you can move fs.writeFileSync(...) to a separate function, but I'm not sure this change improves readability.

Or did you mean something like this:

```js
if (dir doesn't exist or file-in-dir doesn't exist) {
  fs.mkdirSync(...)
  fs.writeFileSync(...)
}
```

However, I'm not sure if fs.mkdirSync() succeeds if the directory already existed; if it fails in that case, this doesn't work.

eginez replied:
yes, I see we'll have to repeat the `writeFile`, no worries then.

eginez commented on docker-compose.yml:
```diff
@@ -71,6 +71,9 @@ services:
     build: test
+    volumes:
+      # We need access to the previous responses and be able to save responses for any new tests.
+      - ./test/saved-responses:/app/saved-responses
```
I am curious about why we are mounting the directory as opposed to copying the files to the docker container.

eginez replied:
If we keep them in the docker container, downloading new test files will be just a matter of adding such files to the remote storage

jan-niestadt replied:
The idea behind the bind mount is that if there was no saved response, the test passes and the response is saved. This means I can add a test, run the script, and commit the test and response right away.

Of course we don't need the bind mount on GitHub, so now there's two scripts: `run-local.sh` that uses the bind mount and saves missing responses, and `run-ci.sh` for GitHub without the bind mount that uses the responses from inside the image (or downloaded in your case) and fails if one is missing.

I'd like to regularly add new tests. If you have any suggestions on how to reduce friction on your side, let me know.

eginez replied:
great that makes sense, thanks for doing that!

eginez commented on test/test/hits-unchanged.js:
```diff
@@ -0,0 +1,74 @@
+
+// Test that a hits search for a pattern returns the correct number of hits and docs,
+// and optionally test that the first hit matches (either JSON or text).
+function expectHitsUnchanged(testName, params) {
```
`testName` is also the test file name right?

jan-niestadt replied:
Correct, added to comments.

eginez reviewed (approved):
I like this change. It will indeed improve testing, this is essentially what I was doing for indices.

>Can this work with your separate run with different test settings as well?

It definitely can, however, it think it would be much easier if we copy the data into the docker container as opposed to mount a new directory to access those files, in a similar fashion as: `INDEX_TEST_DATA_ROOT` (which can probably be rename to TEST_DATA_ROOT)

I notice the file names match the rule pattern, I am not too keen on that, mostly because those patterns can have characters that will  trip up the file system. I would reconsider naming those files to something different and make the file name part of the test. Other than that I left a few minor comments above.

eginez commented:
I think I ran into that issue as well when trying to do something similar to this. I will be glad to have a way to request idempotent responses.
> One issue is that the sort operation is not completely defined, resulting in different ordering of hits. I think I will add an option to sort by (doc, start, end) as the final sort option to fix this.

jan-niestadt commented:
> INDEX_TEST_DATA_ROOT (which can probably be rename to TEST_DATA_ROOT)

Done.

> those patterns can have characters that will trip up the file system

I did sanitize the pattern to produce a valid file name, but I guess you're right. It just means you have to come up with a name for each test which is a bit annoying, but oh well. Done.

> I think I ran into that issue as well when trying to do something similar to this. I will be glad to have a way to request idempotent responses.

Done, I just use `field:pid,hitposition` for the `sort` parameter now.

eginez commented:
> Done, I just use `field:pid,hitposition` for the `sort` parameter now.
Ah I see I did know you could do that. I think I have noticed is that some of the parameters to requests are no immediately obvious, and I also haven't been able to find in them in the documentation. 

A general comment, unrelated to this PR, I think the project would benefit from documenting routes and parameters for the blacklab server.

eginez commented:
Another minor detail I jsut noticed. The location of `saved-responses` is side-by-side with `data` in the `test` directory. Was that intentional? Would it make more sense to move it inside `data`?

jan-niestadt commented:
> I think the project would benefit from documenting routes and parameters for the blacklab server. Maybe this already exists somewhere?

Right now there's only the [BLS overview](https://inl.github.io/BlackLab/blacklab-server-overview.html), but that's confusing and incomplete. The rest of the site was also due for an overhaul.

That's why I've been working on a new site. See the [site dir in the vuepress-site branch](https://github.com/INL/BlackLab/tree/vuepress-site/site). Feedback and PRs welcome. I think we'll switch over pretty soon as it's already an improvement, even if there's more to do.

> The location of saved-responses is side-by-side with data in the test directory. Was that intentional? Would it make more sense to move it inside data?

Agreed, I've moved it inside `data/` now. While I was at it, I've also moved the input files to a subdir `input/` and the `expected-*.json` used by `indices.js` to `data/saved-responses/indices/`. I've also changed how the Docker containers mount the input files and saved responses. See https://github.com/INL/BlackLab/pull/331/commits/a804c91a840e6b624aaef360a8ff928bef165f18 for the details. You will probably need to update your custom tests; hopefully moving the input files to an `input/` subdir there as well is enough.

jan-niestadt commented:
Manually squash merged this to dev.
